proxy/router/src/cache.rs

```diff
@@ -1,88 +1,239 @@
-    vals: IndexMap<K, V>,
+/// - `reserve` computes in O(n) time (average) when capacity is not available,
+pub struct Cache<K: Hash + Eq, V, N: Now = ()> {
+    vals: IndexMap<K, Node<V>>,
     capacity: usize,
+    max_idle_age: Duration,
+
+    /// The time source.
+    now: N,
+}
+
+/// Provides the current time within the module. Useful for testing.
+pub trait Now {
+    fn now(&self) -> Instant;
+}
+
+/// Wraps values
+#[derive(Debug, PartialEq)]
+pub struct Node<T> {
+    value: T,
+    last_access: Instant,
+}
+
+/// A smart pointer that updates an access time when dropped.
+///
+/// Wraps a mutable reference to a `V`-typed value.
+///
+/// When the guard is dropped, the value's `last_access` time is updated with the provided
+/// time source.
+#[derive(Debug)]
+pub struct Access<'a, T: 'a, N: Now + 'a = ()> {
+    node: &'a mut Node<T>,
+    now: &'a N,
 }
 
-pub struct Reserve<'a, K: Hash + Eq + 'a, V: 'a> {
-    vals: &'a mut IndexMap<K, V>,
+#[derive(Debug)]
+pub struct Reserve<'a, K: Hash + Eq + 'a, V: 'a, N: 'a> {
+    vals: &'a mut IndexMap<K, Node<V>>,
+    now: &'a N,
 }
 
 #[derive(Clone, Debug, PartialEq)]
 pub struct CapacityExhausted {
     pub capacity: usize,
 }
 
-impl<K: Hash + Eq, V> Cache<K, V> {
-    pub fn new(capacity: usize) -> Self {
+// ===== impl Cache =====
+
+impl<K: Hash + Eq, V> Cache<K, V, ()> {
+    pub fn new(capacity: usize, max_idle_age: Duration) -> Self {
         Self {
             capacity,
             vals: IndexMap::default(),
+            max_idle_age,
+            now: (),
         }
     }
+}
 
+impl<K: Hash + Eq, V, N: Now> Cache<K, V, N> {
     /// Accesses a route.
-    // TODO track access times for each entry.
-    pub fn access<Q>(&mut self, key: &Q) -> Option<&mut V>
+    ///
+    /// A mutable reference to the route is wrapped in the returned `Access` to
+    /// ensure that the access-time is updated when the reference is released.
+    pub fn access<Q>(&mut self, key: &Q) -> Option<Access<V, N>>
     where
         Q: Hash + Equivalent<K>,
     {
-        self.vals.get_mut(key)
+        let v = self.vals.get_mut(key)?;
+        Some(v.access(&self.now))
     }
 
     /// Ensures that there is capacity to store an additional route.
     ///
+    /// Returns a handle that may be used to store an ite,. If there is no available
+    /// capacity, idle entries may be evicted to create capacity.
+    ///
     /// An error is returned if there is no available capacity.
-    // TODO evict old entries
-    pub fn reserve(&mut self) -> Result<Reserve<K, V>, CapacityExhausted> {
-        let avail = self.capacity - self.vals.len();
-        if avail == 0 {
-            // TODO If the cache is full, evict the oldest inactive route. If all
-            // routes are active, fail the request.
-            return Err(CapacityExhausted {
-                capacity: self.capacity,
-            });
+    pub fn reserve(&mut self) -> Result<Reserve<K, V, N>, CapacityExhausted> {
```

Comment:
This has a bad worse case of a linear scan per request.

The slack discussion conclusion was to not using a doubly linked list to keep entries ordered by access to avoid additional overhead during access.

Another option would be to have a time granularity (like, 1 second), so that many requests within a single second would not result in repeated linear scans that don't do much.

Reply:
Increasing the granularity seems like it might be a good call to me.